cpmpy/transformations/linearize.py

```diff
@@ -5,190 +5,183 @@
 
-        return [cond.implies(l_expr) for l_expr in lin_exprs]
 
-    # Binary operators
-    if cpm_expr.name == "<":
-        # TODO: this can be optimized, see https://github.com/CPMpy/cpmpy/issues/97
-        lhs, rhs = cpm_expr.args
-        if lhs.name == "wsum":
-            return [Operator("wsum", [lhs.args[0] + [1], lhs.args[1] + [1]]) <= rhs]
-        else:
-            cons = lhs + 1 <= rhs
-        cons = flatten_constraint(cons)
-        cons = linearize_constraint(cons)
-        return cons
-
-    if cpm_expr.name == ">":
-        # TODO: this can be optimized, see https://github.com/CPMpy/cpmpy/issues/97
-        lhs, rhs = cpm_expr.args
-        if lhs.name == "wsum":
-            return [Operator("wsum", [lhs.args[0] + [-1], lhs.args[1] + [1]]) >= rhs]
-        else:
-            cons = lhs - 1 >= rhs
-        cons = flatten_constraint(cons)
-        cons = linearize_constraint(cons)
-        return cons
-
-    if cpm_expr.name == "!=":
-        # TODO: this can be optimized, see https://github.com/CPMpy/cpmpy/pull/93/files#r824163315
-        lhs, rhs = cpm_expr.args
-        # Special case: BV != BV
-        if isinstance(lhs, _BoolVarImpl) and isinstance(rhs, _BoolVarImpl):
-            return [lhs + rhs == 1]
-        # Normal case: big M implementation
-        z = boolvar()
-
-        # Calculate bounds of M = |lhs - rhs| + 1
-        bound1, _ = get_or_make_var(1 + lhs - rhs)
-        bound2, _ = get_or_make_var(1 + rhs - lhs)
-        M = max(bound1.ub, bound2.ub)
-
-        if lhs.name == "sum":
-            lhs = Operator("wsum", [[1]*len(lhs.args), lhs.args])
-            cons_lhs = []
-        if _wsum_should(lhs):
-            lhs, cons_lhs = Operator("wsum",_wsum_make(lhs)), []
-        if lhs.name != "wsum":
-            lhs, cons_lhs = get_or_make_var(lhs)
-        # Rewrite of constraints:
-        #   c1 = lhs - rhs <= Mz - 1
-        #   c2 = lhs - rhs >= -M(1-z) + 1
-        c1 = Operator("wsum", [[-M, 1],       [z, 1]]) + lhs <= rhs      # TODO: this can be optimized, see https://github.com/CPMpy/cpmpy/issues/97
-        c2 = Operator("wsum", [[-M, (M - 1)], [z, 1]]) + lhs >= rhs      # TODO: this can be optimized, see https://github.com/CPMpy/cpmpy/issues/97
-
-        return linearize_constraint(flatten_constraint(cons_lhs)) + [c1, c2]
-
-    if cpm_expr.name in [">=", "<=", "=="] and \
-            isinstance(cpm_expr.args[0], Operator) and cpm_expr.args[0].name == "mul":
-        lhs, rhs = cpm_expr.args
-        if len(lhs.args) == 2:
-            # multiplication of var and constant
-            if _wsum_should(lhs):
-                return [Comparison(cpm_expr.name, Operator("wsum", _wsum_make(lhs)), rhs)]
-            return [cpm_expr]
+        # BV -> LinExpr
+        if isinstance(cond, _BoolVarImpl):
+            lin_sub = linearize_constraint(sub_expr, supported=supported)
+            return [cond.implies(lin) for lin in lin_sub]
+        # # LinExpr -> BV Updated: unsupported by new grammar
```

Comment:
it will not appear in the input grammar?
or it can but then wont be supported? Because equiv to ~bv -> ~linexpr, which we could do somewhere

Reply:
Yeah, it can happen and it should be supported. In general, it requires new variables though.

E.g. `x==y -> p` as input should lead to `aux1 or aux2 or p`, `aux1 -> x>y`, `~aux1 -> x<=y`, `aux2 -> x<y`, `~aux2 -> x>=y`, which can all be translated with previous methods.

I am not sure what "new grammar" refers to here. The new grammar we discussed on https://github.com/CPMpy/cpmpy/issues/183#issuecomment-1412363735 is the grammar that linearize outputs, so it does not restrict the input.